test/unit/puppetlabs/puppetserver/lockable_pool_test.clj

```diff
@@ -167,54 +207,314 @@
+            lock-thread-2 (future (deliver lock-thread-started-2? true)
+                                  (.lock pool)
+                                  (deliver lock-thread-locked-2? true)
+                                  (.unlock pool))]
+        @lock-thread-started-1?
+        @lock-thread-started-2?
+        (is (not (realized? lock-thread-locked-1?)))
+        (is (not (realized? lock-thread-locked-2?)))
+        (.unlock pool)
+        (is (true? true))
+        @lock-thread-1
+        (is (true? true))
+        @lock-thread-2
+        (is (true? true))))))
+
+(deftest pool-lock-not-held-after-thread-interrupt
+  (let [pool (create-populated-pool 1)
+        item (.borrowItem pool)
+        lock-thread-obj (promise)
+        lock-thread-locked? (promise)
+        lock-thread (future (deliver lock-thread-obj (Thread/currentThread))
+                            (.lock pool)
+                            (deliver lock-thread-locked? true))]
+    (testing (str "write locker's thread can be interrupted while waiting for "
+                  "instances to be returned")
+      (.interrupt @lock-thread-obj)
+      (is (thrown? ExecutionException @lock-thread))
+      (is (not (realized? lock-thread-locked?))))
+    (.releaseItem pool item)
+    (testing "new write lock can be taken after prior write lock interrupted"
+      (.lock pool)
+      (is (true? true)))))
+
+(deftest pool-unlock-from-thread-not-holding-lock-fails
+  (testing "call to unlock pool when no lock held throws exception"
+    (let [pool (create-populated-pool 1)]
+      (is (thrown? IllegalStateException (.unlock pool)))))
+  (testing "call to unlock pool from thread not holding lock throws exception"
+    (let [pool (create-populated-pool 1)
+          lock-started? (promise)
+          unlock? (promise)
+          lock-thread (future (.lock pool)
+                              (deliver lock-started? true)
+                              @unlock?
+                              (.unlock pool))]
+      @lock-started?
+      (is (thrown? IllegalStateException (.unlock pool)))
+      (deliver unlock? true)
+      @lock-thread
+      (is (true? true)))))
+
 (deftest pool-release-item-test
   (testing (str "releaseItem call with value 'false' does not return item to "
                 "pool but does allow pool to still be lockable")
-    (let [pool (create-populated-pool 1)
-          lock (jruby-testutils/get-lock-from-pool pool)
+    (let [pool (create-populated-pool 2)
           instance (.borrowItem pool)]
-      (is (= 0 (.size pool)))
+      (is (= 1 (.size pool)))
       (.releaseItem pool instance false)
-      (is (= 0 (.size pool)))
+      (.unregister pool instance)
+      (is (= 1 (.size pool)))
       (.lock pool)
-      (is (.isWriteLocked lock))
+      (is (nil? @(future (.borrowItemWithTimeout pool
+                                                 1
+                                                 TimeUnit/MICROSECONDS))))
```

Comment:
> Other thing I was going to mention is that it seems like this pattern of specifying an explicit timeout with the borrow operation might be better than the more global borrow-timeout: 1 that you put in the config of one of the other tests

That makes sense.  I did the `borrow-timeout: 1` setting in the app configuration for the jruby_locking_test tests because those tests go through the JRuby service to borrow instances and there's no function on that service that allows the timeout to be specified per request.  I think for the lockable-pool-tests that I specified a timeout in the `.borrowItemWithTimeout` call directly where a timeout was needed.  I had the ability to do that at this level since these tests use the `JRubyPool` class directly.